### ISSUE TITLE
Sorting bookmarks and Custom sync/download/upload

### ISSUE THREAD

**Issue by simurq** [xBrowserSync/App#28]
I apologize for putting it here but I couldn't find a better place (eg. forum) after browsing your webpage.

I'm also one of the happy users of xBrowserSync after the demise of Xmarks. I really enjoy it and particularly the sync option I often use on PC and two notebooks. Free space for bookmarks is quite reasonable for now unless you add features for syncing history, browser settings, etc. in the future. You're probably extremely busy these days, so please consider the below as issues of minor (if not at all) relevance to what you already plan for the upcoming release. I can live without 'em for now...

I'm one of those users who add bookmarks often but do it in organised manner (read: hierarchy). I seldom use the search function, as I know in which particular folder I added the bookmark(s) and try to keep them sorted alphabetically as often as I can. But since there is no option to track the course of syncing events (eg. progress bar?), it seems that xBrowserSync skips (or not programmed to bother at all?) the order of bookmarks / folder structure between devices even when 'Service' is online on both devices.

Also, I'd like to suggest adding some functionality to custom sync/download/upload of bookmarks BUT asking the user the expected behavior in advance (to keep / erase bookmarks on the server / device) so that one couldn't inadvertently ruin the whole process. Sometimes it may be necessary to switch between the devices and to update bookmarks often; so, custom sync option would come in handy indeed...

Thanks and Good luck!

Rustam

**Comment by nero120** [comment]
Hi @simurq, very happy to hear you're enjoying xBrowserSync, and thank you for the nice comments. 😄 

If I'm understanding you correctly, what you're describing has already been implemented so if it's not working correctly perhaps you've discovered a bug. xBrowserSync is designed to work with Chrome's native bookmarking features, so if (as you say) you don't want to use xBrowserSync's interface then you can simply use the browser's bookmark button, bookmarks toolbar and bookmarks manager and every one of your changes made using those features will be tracked and synced sequentially in the background. For example:

1. Device 1 and device 2 are synced to the same xBrowserSync ID.
2. On device 1, I open Chrome's bookmark manager and drag three bookmarks from other bookmarks to the bookmarks bar (note: I have enabled the xBrowserSync sync setting "Include bookmarks bar" on both devices). After moving the bookmarks, the xBrowserSync icon changes for a short time to indicate a sync is being performed.
3. On device 2, I have not made any bookmarks changes *for five minutes* since making the changes on device 1. The xBrowserSync icon automatically changes to show that a sync is being performed, and the three bookmarks now appear in the bookmarks bar as per device 1.

That's a very simple example, but you should be seeing all of your changes using the browser's native bookmarking features being synced. It is possible that if you are making changes immediately between multiple devices then you will get sync conflicts, which are currently not resolved in any way so you'll get a warning that a conflict was detected and your previous change was not synced.

I hope that makes things clearer. If you believe you are experiencing a bug, could I ask you to provide a detailed list of steps that I can follow to recreate the issue? Also could you just confirm which browser you are using (and it's version) and the version of xBrowserSync you are using?

**Comment by simurq** [comment]
Hi **nero120**, thanks and sorry for late reply.

I make all changes on PC using Chrome v68 native bookmark manager, exactly the same way you've described above (even the "Include bookmarks bar" option is on). Also, xBrowserSync (xBS) uses the **same** ID linked to https://api.xbrowsersync.org both on PC and my notebook. But when I make changes either on PC or on notebook, the app notifies me that it couldn't connect to the xBS sync service  asking me to try next time. I've reinstalled the app a couple of times to no avail yet... I mostly use PC and in fact it's more than a week that the latest changes on PC are not synced with my notebook (still has the old bookmarks). When I check the service status on xBS interface, I can mostly see it offline... Don't know why though cause both of my devices are always online. As if xBS needs a better proof of my online status. :)

I also noticed that when I make any change to bookmarks through xBS and push the save button, it activates instantly and replaces all bookmarks with the "refreshed" version. I cannot see the same behavior when I edit the bookmarks through Chrome's native bookmark manager, as I guess the syncing occurs in "silent" mode.

That's all I can say for now. I hope my answer satisfies your request about a detailed explanation of my existing problem with syncing. And I'm looking forward to hear your comment about custom syncing option in xBS.

Thank you very much!!!

**Comment by nero120** [comment]
@simurq apologies for the delay in getting back to you, lots going on and I lost track of this issue.

Are you still experiencing loss of connectivity to https://api.xbrowsersync.org? That is strange as I have had no downtime alerts (I monitor the service to ensure it's always running), could it possibly be connectivity at your end? If it happens again could you try accessing https://api.xbrowsersync.org on your mobile (using mobile data, not your wifi to ensure a true independent test) and see if you can get to it?

Regarding your original issue, I have recently fixed an issue regarding automatic checking for updates. You can verify this by doing the following:

1. On PC 1, make a change in the Chrome Bookmarks manager (whilst synced)
2. Immediately on PC 2, disable **then re-enable** the sync in xBrowserSync. The bookmarks should be refreshed and your changes made in step 1 should now be visible.

In short, I believe that your changes were indeed synced, but the other machines were not automatically pulling down the updates.

If you could confirm either way I'd really appreciate it!

**Comment by simurq** [comment]
**@nero120,**

Thank you very much for your reply and the updated version! All connectivity problems seem resolved with this new version.

Cheers! And please keep xBrowserSync an ongoing project, we definitely need it!

**Comment by nero120** [comment]
Great news @simurq, and thanks for the kind words! 😊
